local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,227 @@ fn test_slot_hash_expiry() {
+        if last_vote >= common_ancestor_slot {
+            common_ancestor_slot = last_vote;
+            info!("common_ancestor_slot: {}", common_ancestor_slot);
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+        sleep(Duration::from_millis(100));
+    }
+
+    // Keep A running, but setup B so that it thinks it has voted up until common ancestor (but
+    // doesn't know anything past that)
+    {
+        info!("Copying A's ledger to B");
+        std::fs::remove_dir_all(&b_info.info.ledger_path).unwrap();
+        let mut opt = fs_extra::dir::CopyOptions::new();
+        opt.copy_inside = true;
+        fs_extra::dir::copy(&a_ledger_path, &b_ledger_path, &opt).unwrap();
+
+        // remove A's tower in B's new copied ledger
+        info!("Removing A's tower in B's ledger dir");
+        remove_tower(&b_ledger_path, &a_pubkey);
+
+        // Get rid of any slots past common_ancestor_slot
+        info!("Removing extra slots from B's blockstore");
+        let blockstore = open_blockstore(&b_ledger_path);
+        purge_slots_with_count(&blockstore, common_ancestor_slot + 1, 100);
+    }
+
+    info!("Run A on majority fork until it creates its own fork");
+    let mut last_vote_on_a;
+    // Keep A running for a while longer so A has its own fork.
+    let now = Instant::now();
+    loop {
+        last_vote_on_a = wait_for_last_vote_in_tower_to_land_in_ledger(&a_ledger_path, &a_pubkey);
+        if last_vote_on_a > common_ancestor_slot + 3 {
+            let blockstore = open_blockstore(&a_ledger_path);
+            info!(
+                "A majority fork: {:?}",
+                AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>()
+            );
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+        sleep(Duration::from_millis(100));
+    }
+
+    // Kill A and restart B with voting. B should now fork off
+    info!("Killing A");
+    let a_info = cluster.exit_node(&a_pubkey);
+
+    info!("Restarting B");
+    b_info.config.voting_disabled = false;
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    // Allow B to fork now.
+    info!("Allowing B to fork");
+    let now = Instant::now();
+    loop {
+        sleep(Duration::from_millis(200));
+        let (last_vote, _) = last_vote_in_tower(&b_ledger_path, &b_pubkey).unwrap();
+        if last_vote > common_ancestor_slot + 3 {
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+    }
+    let last_vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
```

Comment:
It's because B is killed as soon as next block is generated, so last_vote_on_b hasn't landed yet.

Reply:
hmmm, but wouldn't `B` make another block on top of its own block the next time it's the leader? The `slot_hashes` isn't counting the distance between the blocks, but the *number* of such blocks that are being built on a specific fork

Reply:
Hmm, so the idea is `B`'s vote won't be in gossip on restart because the validator killed before that vote can propagate. I think there's a slight race condition here because validator shut down is slow, might be safe just to also truncate B's ledger past it's last vote

Reply:
ping on this last comment, everything else is looking good!

Reply:
Changed.